src/server/scripts/EasternKingdoms/ZulGurub/boss_thekal.cpp

```diff
@@ -15,443 +15,489 @@
-
-            if (Creature* zealot = instance->GetCreature(DATA_LORKHAN))
-            {
-                zealot->AI()->Reset();
-            }
+            return !me->HasUnitState(UNIT_STATE_CASTING);
+        });
+    }
 
-            if (Creature* zealot = instance->GetCreature(DATA_ZATH))
-            {
-                zealot->AI()->Reset();
-            }
+    void JustDied(Unit* /*killer*/) override
+    {
+        _JustDied();
+        Talk(SAY_DEATH);
 
-            _scheduler.SetValidator([this]
-            {
-                return !me->HasUnitState(UNIT_STATE_CASTING);
-            });
+        if (Creature* zealot = instance->GetCreature(DATA_LORKHAN))
+        {
+            zealot->Kill(zealot, zealot);
         }
 
-        void JustDied(Unit* /*killer*/) override
+        if (Creature* zealot = instance->GetCreature(DATA_ZATH))
         {
-            _JustDied();
-            Talk(SAY_DEATH);
+            zealot->Kill(zealot, zealot);
+        }
+    }
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _JustEngagedWith();
+
+        scheduler.CancelAll();
+        scheduler.Schedule(4s, [this](TaskContext context) {
+            DoCastVictim(SPELL_MORTALCLEAVE);
+            context.Repeat(15s, 20s);
+        }).Schedule(9s, [this](TaskContext context) {
+            DoCastVictim(SPELL_SILENCE);
+            context.Repeat(20s, 25s);
+        }).Schedule(16s, [this](TaskContext context) {
+            DoCastSelf(SPELL_BLOODLUST);
+            context.Repeat(20s, 28s);
+        });
+    }
 
-            if (Creature* zealot = instance->GetCreature(DATA_LORKHAN))
+    void SetData(uint32 /*type*/, uint32 data) override
+    {
+        UpdateZealotStatus(data, true);
+        CheckPhaseTransition();
+
+        scheduler.Schedule(10s, [this, data](TaskContext /*context*/) {
+            if (!_lorkhanDied || !_zathDied || !_wasDead)
             {
-                zealot->Kill(zealot, zealot);
+                ReviveZealot(data);
             }
+        });
+    }
+
+    void DamageTaken(Unit* /*attacker*/, uint32& damage, DamageEffectType, SpellSchoolMask) override
+    {
+        if (!me->HasAura(SPELL_TIGER_FORM) && damage >= me->GetHealth())
+        {
+            damage = me->GetHealth() - 1;
 
-            if (Creature* zealot = instance->GetCreature(DATA_ZATH))
+            if (!_wasDead)
             {
-                zealot->Kill(zealot, zealot);
+                me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
+                me->SetReactState(REACT_PASSIVE);
+                me->SetStandState(UNIT_STAND_STATE_DEAD);
+                me->AttackStop();
+                DoResetThreatList();
+                _wasDead = true;
+                CheckPhaseTransition();
+                Talk(EMOTE_DIES);
             }
         }
 
-        void JustEngagedWith(Unit* /*who*/) override
+        if (!_enraged && me->HealthBelowPctDamaged(20, damage) && me->HasAura(SPELL_TIGER_FORM))
```

Comment:
Schedule health check event on p2 begin?

Reply:
This is leftover from the original script but I don't mind updating it. I'll see if I can figure out how to bake all those conditions into the `ScheduleHealthCheckEvent` call.

Reply:
Finally figured this out, done in [1b405db](https://github.com/azerothcore/azerothcore-wotlk/pull/17603/commits/1b405db2f2e73fd508412d6fa3a42d6dadd8d7f4)